File: backend/app/api/modems.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List
from datetime import datetime, date

from app.core.database import get_db
from app.models.modem import Modem
from app.models.sms import SmsMessage, SmsDirection
from app.schemas.modem import ModemOut, ModemUpdate, ModemDetail
from app.services import modem_manager

router = APIRouter(prefix="/modems", tags=["modems"])


@router.get("/", response_model=List[ModemOut])
def list_modems(db: Session = Depends(get_db)):
    return db.query(Modem).order_by(Modem.id).all()


@router.get("/{modem_id}", response_model=ModemOut)
def get_modem(modem_id: int, db: Session = Depends(get_db)):
    modem = db.query(Modem).filter(Modem.id == modem_id).first()
    if not modem:
        raise HTTPException(status_code=404, detail="Modem not found")
    return modem


@router.patch("/{modem_id}", response_model=ModemOut)
def update_modem(modem_id: int, data: ModemUpdate, db: Session = Depends(get_db)):
    modem = db.query(Modem).filter(Modem.id == modem_id).first()
    if not modem:
        raise HTTPException(status_code=404, detail="Modem not found")
    if data.alias is not None:
        modem.alias = data.alias
    db.commit()
    db.refresh(modem)
    return modem


@router.get("/{modem_id}/detail", response_model=ModemDetail)
def get_modem_detail(modem_id: int, db: Session = Depends(get_db)):
    modem = db.query(Modem).filter(Modem.id == modem_id).first()
    if not modem:
        raise HTTPException(status_code=404, detail="Modem not found")

    sms_sent = db.query(func.count(SmsMessage.id)).filter(
        SmsMessage.modem_id == modem_id,
        SmsMessage.direction == SmsDirection.OUTBOUND,
    ).scalar() or 0

    sms_received = db.query(func.count(SmsMessage.id)).filter(
        SmsMessage.modem_id == modem_id,
        SmsMessage.direction == SmsDirection.INBOUND,
    ).scalar() or 0

    today_start = datetime.combine(date.today(), datetime.min.time())
    sms_today = db.query(func.count(SmsMessage.id)).filter(
        SmsMessage.modem_id == modem_id,
        SmsMessage.created_at >= today_start,
    ).scalar() or 0

    return ModemDetail(
        **{c.name: getattr(modem, c.name) for c in modem.__table__.columns},
        sms_sent=sms_sent,
        sms_received=sms_received,
        sms_today=sms_today,
    )


@router.post("/{modem_id}/refresh", response_model=ModemOut)
def refresh_modem(modem_id: int, db: Session = Depends(get_db)):
    """Force refresh modem info from ModemManager."""
    modem = db.query(Modem).filter(Modem.id == modem_id).first()
    if not modem or not modem.mm_object_path:
        raise HTTPException(status_code=404, detail="Modem not found")
    info = modem_manager.get_modem_info(modem.mm_object_path)
    if not info:
        raise HTTPException(status_code=503, detail="Could not reach modem")
    modem.signal_quality = info.get("signal_quality", 0)
    modem.operator = info.get("operator", "")
    modem.status = info.get("status", "unknown")
    db.commit()
    db.refresh(modem)
    return modem

File: backend-go/handlers/modems.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"simnexus-go/database"
	"simnexus-go/middleware"
	"simnexus-go/models"
	"simnexus-go/security"
	"simnexus-go/services"

	"github.com/gin-gonic/gin"
)

// visibleModemIDs returns (ids, unrestricted, permitted).
// unrestricted=true means admin/no filter. permitted=false means 403.
func visibleModemIDs(u *models.User) ([]uint, bool, bool) {
	if u.Role == models.RoleAdmin {
		return nil, true, true
	}
	p := security.Perm(u)
	if p == nil || !p.CanViewSim {
		return nil, false, false
	}
	granted := security.GetUserModemGrants(database.DB, u.ID, "", u)
	// intersect with restricted role scope (non-approver roles only)
	if !p.CanApproveRequests && p.AllowedModemIDs != nil {
		filtered := granted[:0]
		for _, id := range granted {
			if security.ContainsUint(p.AllowedModemIDs, id) {
				filtered = append(filtered, id)
			}
		}
		granted = filtered
	}
	return granted, false, true
}

// ListAvailableModems returns all active modems for browsing.
func ListAvailableModems(c *gin.Context) {
	u := middleware.CurrentUser(c)
	if u.Role != models.RoleAdmin {
		p := security.Perm(u)
		if p == nil || !p.CanViewSim {
			c.JSON(http.StatusForbidden, gin.H{"detail": "无SIM卡查看权限"})
			return
		}
	}
	var modems []models.Modem
	database.DB.Where("is_active = ?", true).Order("id").Find(&modems)
	c.JSON(http.StatusOK, modems)
}

// ListModems returns modems visible to the user.
func ListModems(c *gin.Context) {
	u := middleware.CurrentUser(c)
	if u.Role == models.RoleAdmin {
		var modems []models.Modem
		database.DB.Where("is_active = ?", true).Order("id").Find(&modems)
		c.JSON(http.StatusOK, modems)
		return
	}
	ids, _, ok := visibleModemIDs(u)
	if !ok {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无SIM卡查看权限"})
		return
	}
	if len(ids) == 0 {
		c.JSON(http.StatusOK, []models.Modem{})
		return
	}
	var modems []models.Modem
	database.DB.Where("id IN ? AND is_active = ?", ids, true).Order("id").Find(&modems)
	c.JSON(http.StatusOK, modems)
}

func canAccessModem(u *models.User, modemID uint) bool {
	if u.Role == models.RoleAdmin {
		return true
	}
	ids, _, ok := visibleModemIDs(u)
	if !ok {
		return false
	}
	return security.ContainsUint(ids, modemID)
}

// GetModem returns a single modem.
func GetModem(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	u := middleware.CurrentUser(c)
	if u.Role != models.RoleAdmin && !canAccessModem(u, uint(id)) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权访问该设备"})
		return
	}
	var modem models.Modem
	if database.DB.First(&modem, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Modem not found"})
		return
	}
	c.JSON(http.StatusOK, modem)
}

type modemUpdate struct {
	Alias *string `json:"alias"`
}

// UpdateModem sets the modem alias.
func UpdateModem(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	var modem models.Modem
	if database.DB.First(&modem, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Modem not found"})
		return
	}
	var data modemUpdate
	c.ShouldBindJSON(&data)
	if data.Alias != nil {
		modem.Alias = *data.Alias
	}
	database.DB.Save(&modem)
	c.JSON(http.StatusOK, modem)
}

// GetModemDetail returns a modem plus SMS stats.
func GetModemDetail(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	u := middleware.CurrentUser(c)
	if u.Role != models.RoleAdmin && !canAccessModem(u, uint(id)) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权访问该设备"})
		return
	}
	var modem models.Modem
	if database.DB.First(&modem, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Modem not found"})
		return
	}
	var sent, received, today int64
	database.DB.Model(&models.SmsMessage{}).Where("modem_id = ? AND direction = ?", id, models.SmsOutbound).Count(&sent)
	database.DB.Model(&models.SmsMessage{}).Where("modem_id = ? AND direction = ?", id, models.SmsInbound).Count(&received)
	todayStart := time.Now().Truncate(24 * time.Hour)
	database.DB.Model(&models.SmsMessage{}).Where("modem_id = ? AND created_at >= ?", id, todayStart).Count(&today)

	out := gin.H{}
	remarshal(modem, &out)
	out["sms_sent"] = sent
	out["sms_received"] = received
	out["sms_today"] = today
	c.JSON(http.StatusOK, out)
}

// RefreshModem re-reads modem info from the device.
func RefreshModem(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	var modem models.Modem
	if database.DB.First(&modem, id).Error != nil || modem.MmObjectPath == "" {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Modem not found"})
		return
	}
	info := services.GetModemInfo(modem.MmObjectPath)
	if info == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"detail": "Could not reach modem"})
		return
	}
	modem.SignalQuality = info.SignalQuality
	modem.Operator = info.Operator
	modem.Status = info.Status
	database.DB.Save(&modem)
	c.JSON(http.StatusOK, modem)
}
